crates/env/src/engine/mod.rs

```diff
@@ -42,65 +51,117 @@ cfg_if! {
-
 // We only use this function when 1) compiling to Wasm 2) compiling for tests.
 #[cfg_attr(all(feature = "std", not(test)), allow(dead_code))]
-pub(crate) fn decode_fallible_constructor_reverted_return_value<I, E, ContractError>(
+pub(crate) fn decode_instantiate_result<I, E, R, ContractStorage, ContractRef>(
+    instantiate_result: EnvResult<()>,
+    out_address: &mut I,
     out_return_value: &mut I,
-) -> EnvResult<ConstructorResult<ContractResult<E::AccountId, ContractError>>>
+) -> EnvResult<
+    ConstructorResult<<R as InstantiateResult<ContractStorage>>::Output<ContractRef>>,
+>
 where
     I: scale::Input,
     E: crate::Environment,
-    ContractError: scale::Decode,
+    R: InstantiateResult<ContractStorage>,
+    ContractRef: FromAccountId<E>,
 {
-    let out: ConstructorResult<Result<(), ContractError>> =
-        scale::Decode::decode(out_return_value)?;
-
-    match out {
-        ConstructorResult::Ok(ContractResult::Ok(())) => {
-            // Since the contract reverted we don't expect an `Ok` return value from the
-            // constructor, otherwise we'd be in the `AccountId` decoding branch.
-            //
-            // While we could handle this more gracefully, e.g through a `LangError`, we're going to
-            // be defensive for now and trap.
-            panic!(
-                "The callee reverted, but did not encode an error in the output buffer."
-            )
+    match instantiate_result {
+        Ok(()) => {
+            let account_id = scale::Decode::decode(out_address)?;
+            let contract_ref =
+                <ContractRef as FromAccountId<E>>::from_account_id(account_id);
+            let output = <R as InstantiateResult<ContractStorage>>::ok(contract_ref);
+            Ok(Ok(output))
         }
-        ConstructorResult::Ok(ContractResult::Err(contract_error)) => {
-            Ok(ConstructorResult::Ok(ContractResult::Err(contract_error)))
+        Err(EnvError::CalleeReverted) => {
+            let constructor_result_variant = out_return_value.read_byte()?;
+            match constructor_result_variant {
+                // 0 == `ConstructorResult::Ok` variant
+                0 => {
+                    if <R as InstantiateResult<ContractStorage>>::IS_RESULT {
+                        let result_variant = out_return_value.read_byte()?;
+                        match result_variant {
+                            // 0 == `Ok` variant
+                            0 => panic!("The callee reverted, but did not encode an error in the output buffer."),
+                            // 1 == `Err` variant
+                            1 => {
+                                let contract_err = <<R as InstantiateResult<ContractStorage>>::Error
+                                    as scale::Decode>::decode(out_return_value)?;
+                                let err = <R as InstantiateResult<ContractStorage>>::err(contract_err);
+                                Ok(Ok(err))
```

Comment:
nit: I'd use the typedefs here to make it more clear as to which `Result` is which

 ```suggestion
                                EnvResult::Ok(ConstructorResult::Ok(err))
```